### DISCUSSION TITLE
Migrate project webpack 4 to 5 : ProvidePlugin issue ?

### DISCUSSION THREAD

**Discussion by msev** [webpack/webpack#15210] [Q&A]
Hi,

I am migrating a web project from webpack 4 to webpack 5.

In console, i have the following error :

`rivets is not defined`

In webpack 4, without ProvidePlugin, the development output is : 

`eval("(function () {\n  var $document,\n      Cart,\n      CartJS,\n      FORMAT_MONEY_WARNING,\n`

In webpack 4, when i add ProvidePlugin : 

```
new webpack.ProvidePlugin({
  $: 'jquery',
  jQuery: 'jquery',
  rivets: 'rivets',
});
```
Everything is working and development output is now with the wrapper  :

`eval("/* WEBPACK VAR INJECTION */(function(jQuery, rivets) {(function () {\n  var $document,\n      Cart,\n      CartJS,\n      FORMAT_MONEY_WARNING,\n`

But now in webpack 5, the output is always without the wrapper `(function(jQuery, rivets) {` (even with ProvidePlugin) :

`eval("(function() {\n  var $document, Cart, CartJS, FORMAT_MONEY_WARNING`

Am i missing something ? 

Thanks.
Best regards.

**Comment by alexander-akait** [comment]
Can you provide example of the problem?

**Reply by msev** [reply]
In webpack 4, there is a function name `variableInjectionFunctionWrapperStartCode` in this core file : https://github.com/webpack/webpack/blob/v4.46.0/lib/JavascriptGenerator.js#L179

This function is wrapping module when it's necessary.

Is there an equivalent in webpack 5 ?

**Reply by alexander-akait** [reply]
Sorry, I don't understand your problem, why you need wrapper?

**Reply by msev** [reply]
(Sorry, this is my bad english).

I need to retrieve this wrapping `(function(jQuery, rivets) {` around of module CartJS.

To achieve that in webpack 4 `ProvidePlug` works perfectly :

```
new webpack.ProvidePlugin({
  $: 'jquery',
  jQuery: 'jquery',
  rivets: 'rivets',
});
```
output is :
`eval("/* WEBPACK VAR INJECTION */(function(jQuery, rivets) {(function () {\n var $document,\n Cart,\n CartJS,\n FORMAT_MONEY_WARNING,\n`

But in webpack 5 (with the same ProvidePlugin configuration), the output is not the same as webpack 4, output is : 

`eval("(function() {\n var $document, Cart, CartJS, FORMAT_MONEY_WARNING`

Without this wrapping, an error is printing in console : `rivets in not defined`

**Reply by alexander-akait** [reply]
Please create reproducible example

**Reply by msev** [reply]
Problem solved.

Add a new rule do the trick :

```
module: {
    rules: [
      {
        test: require.resolve('shopify-cartjs'),
        use: 'imports-loader?rivets'
      }
    ]
  },
```

Sorry for the inconvenience.
Best regards
